lib/service_actor/collectionable.rb

```diff
@@ -6,26 +6,66 @@
 # Example:
 #
 #   class Pay < Actor
-#     input :provider, in: ["MANGOPAY", "PayPal", "Stripe"]
+#     input :provider, inclusion: ["MANGOPAY", "PayPal", "Stripe"]
+#   end
+#
+#   class Pay < Actor
+#     input :provider,
+#           inclusion: {
+#             in: ["MANGOPAY", "PayPal", "Stripe"],
+#             message: (lambda do |_input_key, value, _inclusion_in|
+#               "Payment system \"#{value}\" is not supported"
+#             end)
+#           }
 #   end
 module ServiceActor::Collectionable
   def self.included(base)
     base.prepend(PrependedMethods)
   end
 
   module PrependedMethods
-    def _call
+    def _call # rubocop:disable Metrics/MethodLength
       self.class.inputs.each do |key, options|
-        next unless options[:in]
+        value = result[key]
+        inclusion = options[:inclusion]
+
+        base_arguments = {
+          input_key: key,
+          value: value
+        }
+
+        inclusion_in, message =
+          define_inclusion_with(inclusion, **base_arguments)
 
-        next if options[:in].include?(result[key])
+        next if inclusion_in.nil?
+        next if inclusion_in.include?(value)
 
-        raise ArgumentError,
-              "Input #{key} must be included in #{options[:in].inspect} " \
-              "but instead was #{result[key].inspect}"
+        raise_error_with(
+          message,
+          **base_arguments,
+          inclusion_in: inclusion_in,
+        )
       end
 
       super
     end
+
+    private
+
+    def define_inclusion_with(inclusion, input_key:, value:)
+      if inclusion.is_a?(Hash) # advanced mode
+        inclusion_in, message = inclusion.values_at(:in, :message)
+      else
+        inclusion_in = inclusion
+        message = "Input #{input_key} must be included " \
+                  "in #{inclusion_in.inspect} but instead " \
+                  "was #{value.inspect}"
```

Comment:
Perhaps we could reduce instanciations by setting this to a private constant lambda. E.g.:

```rb
message = DEFAULT_MESSAGE
```

With:

```rb
DEFAULT_MESSAGE = lambda do |input_key: inclusion_in:, value:|
  "Input #{input_key} must be included in …"
end
private_constant :DEFAULT_MESSAGE
```

Reply:
Do you mean only in this place or for all messages?

Reply:
I think this could actually help for every input 🙏🏻

Reply:
Okay. I'll look at it tomorrow with a fresh mind.

Reply:
Done: https://github.com/sunny/actor/pull/82/commits/d0db4785876a022c8ae47d94bdc80449a4d2ccf7

@sunny Tell me, please, did I understand everything correctly?

Reply:
Yes, right on track! I’ve added a few suggestions. I think what’s missing is using named arguments instead of positional ones.